docs/04-reference/winglang-plugins.md

```diff
@@ -0,0 +1,76 @@
+---
+title: Compiler Plugins
+id: compiler-plugins
+description: Wing Compiler Plugins Reference
+keywords: [Wing reference, Wing language, language, Wing language spec, Wing programming language, compiler plugins]
+---
+
+Wing compiler plugins are used to extend the Wing compiler. They can be used to add non-functional customizations to the compilation's resulting infrastructure. For example, you can use a compiler plugin to enforce a tagging convention on all resources in the resulting infrastructure.
+
+## Plugin Design
+
+A compiler plugin is a JavaScript file that exports multiple functions. These functions are called by the Wing compiler at various points during compilation also referred to as stages.
+
+At each stage the compiler plugin is provided with a context object that allows the plugin to hook into the compilation process and apply customizations.
+
+A single plugin can implement any number of stages. However, it is recommended to only implement the stages that are needed.
+
+## Compilation Stages
+
+Each compilation stage is meant to serve a specific purpose. The following is a list of the stages and their purpose:
+
+### `preSynth` stage
+
+This stage is called before the Wing begins to synthesize. In the context of a cdktf target, this stage will have access to the root node of the construct tree. Allowing for the plugin to add or manipulate constructs before the synthesis process begins.
+
+The following example adds bucket to root node, note this plugin is intended to be used with the `tf-aws` target.
+```js
+const s3_bucket = require("@cdktf/provider-aws/lib/s3-bucket");
+
+// exports the preSynth function
+exports.preSynth = function(app) {
+  // app is the root node of the construct tree
+  new s3_bucket.S3Bucket(app, "MyPluginBucket", {
+    bucket: "my-plugin-bucket",
+  });
+}
+```
+
+### `postSynth` stage
+
+This stage is called after the Wing has finished synthesizing. In the context of a cdktf target, this stage will have access to the raw terraform JSON configuration. Allowing for manipulation of the JSON that is written to the compiled output directory.
+
+This stage is useful for adding customizations that can not be applied in the context of the preSynth phase. Its worth noting that this is not meant as a validation phase since this stage is mutable by subsequent plugins.
+
+The following example manipulates the terraform configuration to use a s3 backend. For brevity, the example uses hard coded values. In practice these would be configurable.
```

Comment:
```suggestion
The following example manipulates the terraform configuration to use a S3 backend. For brevity, the example uses hard coded values. In practice these would be configurable.
```

Reply:
"In practice these would be configurable" like through an environment variable, or something else? (just checking if I understand)

Reply:
Yea I think that configuration in this first iteration of plugins is up to the plugin's creator. So in most cases I would imagine an environment variable, but in the RFC one of the potentials is like a class that can be instantiated from your own plugin with props passed in. Maybe even a config file exists. I added some of those as possibilities for clarity maybe.